lazy-pages/src/sys.rs

```diff
@@ -107,6 +113,191 @@ pub(crate) enum AccessedPagesInfo {
+            .ok()?;
+
+        log::trace!("Change global {name}: {current_value} -> {new_value}, exceeded: {exceed}");
+
+        Some(exceed)
+    };
+    if sub_global(global_gas, amount).ok_or(Error::CannotChargeGas)? {
+        return Ok(Status::GasLimitExceeded);
+    }
+    if sub_global(global_allowance, amount).ok_or(Error::CannotChargeGasAllowance)? {
+        return Ok(Status::GasAllowanceExceeded);
+    }
+    Ok(Status::Normal)
+}
+
+unsafe fn charge_gas(globals_ctx: GlobalsCtx, amount: u64) -> Result<Status, Error> {
+    match globals_ctx.globals_access_mod {
+        GlobalsAccessMod::WasmRuntime => {
+            let instance = (globals_ctx.globals_access_ptr as *mut SandboxInstance)
+                .as_mut()
+                .ok_or(Error::HostInstancePointerIsInvalid)?;
+            charge_gas_internal(
+                GlobalsAccessWasmRuntime { instance },
+                &globals_ctx.global_gas_name,
+                &globals_ctx.global_allowance_name,
+                amount,
+            )
+        }
+        GlobalsAccessMod::NativeRuntime => {
+            let inner_access_provider = (globals_ctx.globals_access_ptr
+                as *mut &mut dyn GlobalsAccesser)
+                .as_mut()
+                .ok_or(Error::DynGlobalsAccessPointerIsInvalid)?;
+            charge_gas_internal(
+                GlobalsAccessNativeRuntime {
+                    inner_access_provider,
+                },
+                &globals_ctx.global_gas_name,
+                &globals_ctx.global_allowance_name,
+                amount,
+            )
+        }
+    }
+}
+
+fn cost_for_write(
+    ctx: &mut RefMut<LazyPagesExecutionContext>,
+    weights: &LazyPagesWeights,
+    page: GranularityPage,
+) -> Result<u64, Error> {
+    if ctx.write_charged.contains(&page) {
+        // Has been already charged for write.
+        Ok(0)
+    } else if ctx.read_charged.contains(&page) {
+        // Has been already charged for read.
+        if ctx.write_after_read_charged.contains(&page) {
+            // Has been already charged for write after read.
+            Ok(0)
+        } else {
+            // Charge for write after read.
+            ctx.write_after_read_charged.insert(page);
+            Ok(weights.write_after_read)
+        }
+    } else if ctx.write_after_read_charged.contains(&page) {
+        Err(Error::WriteAfterReadChargedWithoutReadCharged)
+    } else {
+        // Charge for write.
+        ctx.write_charged.insert(page);
+        Ok(weights.write)
+    }
+}
+
+fn cost_for_read(
+    ctx: &mut RefMut<LazyPagesExecutionContext>,
+    weights: &LazyPagesWeights,
+    page: GranularityPage,
+) -> Result<u64, Error> {
+    if ctx.read_charged.contains(&page) || ctx.write_charged.contains(&page) {
+        // Has been already charged for write or read - so no need to charge for read.
+        Ok(0)
+    } else if ctx.write_after_read_charged.contains(&page) {
+        Err(Error::WriteAfterReadChargedWithoutReadCharged)
```

Comment:
Do we really need this assertion every time?

Reply:
@breathx Actually I'm not sure whether it's really necessary... We also can add infinite number of checks like this)) but something tell me that this one can help as one day